Pegasus/sources/Geometry.cpp

```diff
@@ -212,6 +215,265 @@ bool intersection::CalculateRayAabbIntersection(double tMin, double tMax)
+ */
+intersection::gjk::NearestSimplexData NearestSimplexLineSegment(std::array<glm::dvec3, 4>& simplex)
+{
+    glm::dvec3 const AB = simplex[0] - simplex[1];
+    glm::dvec3 const A0 = glm::dvec3{0, 0, 0} - simplex[1];
+
+    if (intersection::IsSameDirection(AB, A0))
+    {
+        glm::dvec3 const direction = glm::cross(glm::cross(AB, A0), AB);
+        return {2, direction};
+    }
+
+    simplex[0] = simplex[1];
+    return {1, A0};
+}
+
+/**
+ * @brief Finds nearest simplex from the triangle simplex to the origin and returns its size and new search direction
+ * @param[in] simplex triangle simplex vertices
+ * @return new simplex size and search direction
+ */
+intersection::gjk::NearestSimplexData NearestSimplexTriangle(std::array<glm::dvec3, 4>& simplex)
+{
+    glm::dvec3 const AB = simplex[1] - simplex[2];
+    glm::dvec3 const AC = simplex[0] - simplex[2];
+    glm::dvec3 const A0 = glm::dvec3{0, 0, 0} - simplex[2];
+    glm::dvec3 const ABC = glm::cross(AB, AC);
+    intersection::gjk::NearestSimplexData result;
+
+    if (intersection::IsSameDirection(glm::cross(ABC, AC), -simplex[2]))
+    {
+        if (intersection::IsSameDirection(AC, A0))
+        {
+            simplex = {simplex[0], simplex[2]};
+            glm::dvec3 const direction = glm::cross(glm::cross(AC, -simplex[1]), AC);
+            result = {2, direction};
+        }
+        else if (intersection::IsSameDirection(AB, A0))
+        {
+            simplex = {simplex[1], simplex[2]};
+            glm::dvec3 const direction = glm::cross(glm::cross(AB, -simplex[1]), AB);
+            result = {2, direction};
+        }
+        else
+        {
+            simplex = {simplex[2]};
+            glm::dvec3 const direction = -simplex[0];
+            result = {1, direction};
+        }
+    }
+    else if (intersection::IsSameDirection(glm::cross(AB, ABC), -simplex[2]))
+    {
+        if (intersection::IsSameDirection(AB, A0))
+        {
+            simplex = {simplex[1], simplex[2]};
+            glm::dvec3 const direction = glm::cross(glm::cross(AB, -simplex[1]), AB);
+            result = {2, direction};
+        }
+        else
+        {
+            simplex = {simplex[2]};
+            glm::dvec3 const direction = -simplex[0];
+            result = {1, direction};
+        }
+    }
+    else
+    {
+        if (intersection::IsSameDirection(ABC, A0))
+        {
+            glm::dvec3 const direction = ABC;
+            result = {3, direction};
+        }
+        else
+        {
+            glm::dvec3 const direction = -ABC;
+            result = {3, direction};
+        }
+    }
+
+    return result;
+}
+
+/**
+ * @brief Finds nearest simplex from the tetrahedron simplex to the origin and returns its size and new search direction
+ * @param[in] simplex tetrahedron simplex vertices
+ * @return new simplex size and search direction
+ */
+intersection::gjk::NearestSimplexData NearestSimplexTetrahedron(std::array<glm::dvec3, 4>& simplex)
+{
```

Comment:
It would probably be a good idea to do so. Though I'd rather see at the moment a working example. Then the most optimal one. Simply a matter of choice. We'll work on it a bit further on the schedule.